### PR TITLE
Improve documentation for helm upgrade (#4030)

### PR DESCRIPTION
**Pull request by smurfralf**
Expand and clarify documentation for `helm upgrade` to include nuances
of command line values setting.

Fixes issue #4030

### CODE REVIEW

File: cmd/helm/upgrade.go [hunk 1]
```diff
@@ -29,30 +29,57 @@ import (
...
+ - a path to a chart directory,
+ - a packaged chart,
+ - a fully qualified URL.
 
-To override values in a chart, use either the '--values' flag and pass in a file
-or use the '--set' flag and pass configuration from the command line. To force string
-values in '--set', use '--set-string' instead. In case a value is large and therefore
-you want not to use neither '--values' nor '--set', use '--set-file' to read the
-single large value from file.
+To customize the chart values use any of
```
Comment (bacongobbler) [inline cmd/helm/upgrade.go]: can we add the comma back here?

> To customize the chart values, use any of

File: cmd/helm/upgrade.go [hunk 2]
```diff
@@ -29,30 +29,57 @@ import (
...
+ - '--values'/'-f' to pass in a yaml file holding settings,
+ - '--set' to provide one or more key=val pairs directly,
+ - '--set-string' to provide key=val forcing val to be stored as a string,
+ - '--set-file' to provide key=path to read a single large value from a file at path.
 
-You can specify the '--values'/'-f' flag multiple times. The priority will be given to the
-last (right-most) file specified. For example, if both myvalues.yaml and override.yaml
-contained a key called 'Test', the value set in override.yaml would take precedence:
+To edit or append to the existing customized values add the
```
Comment (bacongobbler) [inline cmd/helm/upgrade.go]: > To edit or append to the existing customized values, add the

File: cmd/helm/upgrade.go [hunk 3]
```diff
@@ -29,30 +29,57 @@ import (
...
+ - '--set-file' to provide key=path to read a single large value from a file at path.
 
-You can specify the '--values'/'-f' flag multiple times. The priority will be given to the
-last (right-most) file specified. For example, if both myvalues.yaml and override.yaml
-contained a key called 'Test', the value set in override.yaml would take precedence:
+To edit or append to the existing customized values add the 
+ '--reuse-values' flag, otherwise any existing customized values are ignored.
+
+If no chart value arguments are provided on the command line, any existing customized values are carried
+forward. To revert to use only values provided in the chart, use the '--reset-values' flag.
```
Comment (bacongobbler) [inline cmd/helm/upgrade.go]: This sentence reads a little awkward. Can we change this to

> If you want to revert back to just the values provided in the chart, use the '--reset-values' flag.

File: cmd/helm/upgrade.go [hunk 4]
```diff
@@ -29,30 +29,57 @@ import (
...
-You can specify the '--set' flag multiple times. The priority will be given to the
-last (right-most) set specified. For example, if both 'bar' and 'newbar' values are
-set for a key called 'foo', the 'newbar' value would take precedence:
+Note that the key name provided to the '--set', '--set-string' and '--set-file' flags can reference
+structure elements. Examples:
+  - mybool=TRUE
+  - livenessProbe.timeoutSeconds=10
+  - metrics.annotations[0]=hey,metrics.annotations[1]=ho
+
+which sets the top level key mybool to true, the nested timeoutSeconds to 10 and two array values respectively.
```
Comment (bacongobbler) [inline cmd/helm/upgrade.go]: > the nested timeoutSeconds to 10, and two array values, respectively.

**Review by bacongobbler — CHANGES_REQUESTED** [review]

File: cmd/helm/upgrade.go (hunk 1 above)
Comment (smurfralf) [reply cmd/helm/upgrade.go]: done

File: cmd/helm/upgrade.go (hunk 2 above)
Comment (smurfralf) [reply cmd/helm/upgrade.go]: done

File: cmd/helm/upgrade.go (hunk 3 above)
Comment (smurfralf) [reply cmd/helm/upgrade.go]: done

File: cmd/helm/upgrade.go (hunk 4 above)
Comment (smurfralf) [reply cmd/helm/upgrade.go]: done

**Review by bacongobbler — APPROVED** [review]

**Comment by bacongobbler** [comment]
thanks!

**Pull request merged**